### DISCUSSION TITLE
Failover and replication for EdgeDB

### DISCUSSION THREAD

**Discussion by tailhook** [geldata/gel#1859] [Pre-RFC]
Abstract
========

This proposal describes support of failover and replication for EdgeDB.  Currently this is a collection of things that has to be done, some of them can probably be done outright and some will be integrated into a proper RFC after discussion.

Client-Side Implementation
==========================

Most important work is integrated into [RFC1004](https://github.com/edgedb/rfcs/blob/master/text/1004-transactions-api.rst). Following are missing parts.

Name Resolution
---------------

* Client resolves EdgeDB address by DNS.
* DNS record is refreshed every time new connection happens. I.e. we ignore DNS cache expiration (ensuring that upstream DNS server refreshes at the right time is out of scope for this RFC)
* Established connections are not refreshed, because it's not possible to reliably determine whether new IP address resolved by DNS is because of failover or it's just DNS round-robin returned a different IP.

*(This generally summarizes how it currently works)*

Initial Connect
---------------

Currently when client initially connects to the database and database is inaccessible it crashes. This imposes few issues:

1. If application tries to connect when EdgeDB is being restarted application is likely to crash at least current request or at all
2. `docker-compose up` may not work if it starts EdgeDB and application simultaneously (which is the common case)

To fix this we should have a connection parameter:

```python
edgedb.connect(..., wait_until_available_sec=30)
await edgedb.async_connect(..., wait_until_available_sec=30)
await edgedb.async_connection_pool(..., wait_until_available_sec=30)
```

This method retry reconnecting on the following conditions:

1. Name resolution failed
2. File not found (Unix socket not bound yet)
3. Connection reset, connection aborted, connection refused (server is restarting or not ready yet)
4. Timeout happened

Reconnect should happen when `socket.connect()` call fails or any time during authentication (Docker tends to allow opening a connection and then resetting it afterwards if contained app is not accepting)

Read-Only Requests and Replicas
-------------------------------

Assumptions:

1. Connections to EdgeDB and connections to Postgres are mapped in N:M fasion.  Which means we can send read-only and read-write requests through the same underlying connection and EdgeDB will figure out upstream by itself.
2. Read-Only requests might need to go to primary, but this is minority use case

API looks like:
```python
ro_conn = conn.read_only()
ro_pool = pool.read_only()
ro_conn = conn.read_only(primary=True)
ro_pool = pool.read_only(primary=True)
```

That is:
1. Read-only is allowed to go to replica by default, exact settings of that are determined by initial connection/pool params and can be overriden by `read_only(**params)` call.
2. `primary=False` is default
3. Both for sync and async API `read_only(..)` method returns lightweight object that refers to the same underlying connection/pool object (no await needed)
4. User may store the referred object somewhere or invoke ``read_only`` method
every time as it's cheap
5. Reconfiguring is also possible as in
`pool.read_only().read_only(primary=True)` and the method returns a distinct
lightweight object

See Protocol section for the description of how this works

Also we add a set of replica hosts to the API:
```python
conn = edgedb.connect('myapp_readonly', read_only=True)
cfg = (
    PoolConfig('myapp')
    .max_connections(3)
    .with_readonly(PoolConfig('myapp_readonly').max_connections(10))
)
pool = await edgedb.async_connection_pool(cfg)
```

This is needed to account read-only EdgeDB instances (see below)

Note:

1. Connection can be either read-write or read-only
2. Pool contains both kinds of connections and choses a connection for each
request separately. That is:
```python
    pool = await edgedb.async_connection_pool(cfg)
    ro_pool = pool.read_only()
    await ro_pool.query("...")  # goes to replica
    await ro_pool.read_only(primary=True).query("...")  # goes to primary
    ro2 = ro_pool.read_only(maximum_lag=lag)
    await ro2.query("...")  # goes to replica if that doesn't lag too much
```

(`maximum_lag` parameter may not be implemented in the first iteration)

Retrying Non-trasactional Requests
----------------------------------

After [RFC1004](https://github.com/edgedb/rfcs/blob/master/text/1004-transactions-api.rst) is implemented all transactions are retried by default. But what if error happens in non-trasactional case.

What we aim to achieve is:
```python
    def handler(req, db):
        await db.query(...)
```
Working reliably. Under the hood every query works as follows:

1. `Prepare` message is sent to compile the query
2. `Describe` message is sent to find out the structure of the query
3. `Execute` message is sent to execute the query

This proposal specifies that:

1. `Prepare` should return a flag `READ_ONLY` of `true` or `false` that says
whether query is read-only
2. If query is read-only we retry it in the same manner as `retry` method does
in RFC1004.
3. If query is not readonly we retry only `Prepare` and `Describe` messages and
propagate exception if connection is failed on `Execute` statement.

This means that in most cases users want to use direct `query` on connection (or pool) objects to execute read-only queries and use `retry` and transaction API for all read-write queries. To achieve that we add a knob:

```python
    conn = edgedb.connection(warn={'non_transactional_writes'})
    conn = await edgedb.async_connection(warn={'non_transactional_writes'})
    conn = await edgedb.async_connection_pool(warn={'non_transactional_writes'})
    conn = edgedb.connection(deny={'non_transactional_writes'})
    conn = await edgedb.async_connection(deny={'non_transactional_writes'})
    conn = await edgedb.async_connection_pool(deny={'non_transactional_writes'})
```

* `warn` may be used to trace and update the application (warning will be
emitted by the warning facility of the language)
* `deny` is used to throw errors when `Prepare` infers that query is
non-readonly and query has been done in non-transactional context (not in
`retry` or `try_transaction`)

Note: `warn` and `deny` are supposed to be used for other future warnings too.

Server-Side Implementation
==========================

Faiover of Postgres
-------------------

At stage 1 we assume that if one needs failover they should set up server as
follows:

1. Install Postgres with replication and failover of primary, perhaps by using Amazon RDS or similar service
2. Make sure that Postgres failover is manifested as a change of DNS record or IP failover
3. Make sure that when Postgres master switch all connections to the old ones are dropped (i.e. old instance don't continue to execute queries without reconnect)

At stage 2 we add more logic to EdgeDB:

1. EdgeDB has all the hosts of the instances in the config
2. EdgeDB polls all the Postgres instances to discover a master
3. We provide API in EdgeDB to notify when master changes

And stage 3 is the most complex one:

1. EdgeDB does election of the primary
2. EdgeDB does failover of the primary (monitors and switches replicas)

Stage 1 is implemented preliminarly as follows:

1. In the first implementation it should probably has the same logic as client: resolve DNS name on each connect attempt (without internal cache).
2. Once we implement connection pooling in EdgeDB we should ensure that Postgres master switch is propagated quickly (i.e. client doesn't attempt fail every connection in the pool in sequence until pool eventually creates a new connection)

Replica connections of Postgres
-------------------------------

This proposal specifies that EdgeDB connections to Postgres connections are connected into `M:N` fasion. I.e. the actual connection is picked up from connection pool when either `START TRANSACTION` is executed for the duration of the transaction, or when for any other query for the duration of the query.

This doesn't have to be implemented like this since day one. But this has
the following advantages:

1. Connections in Postgres are costly, this approach should decrease idle time of each connection so requiring less of connections
2. `REPLICA_CONFIG` for each query is flexible. Some queries are okay on lagged replicas. Some queries require reasonably up to date data and this information can be tracked in EdgeDB server much easier and EdgeDB can send request to the primary seamlessly for the client.

So how I see the implementation:
1. EdgeDB has a connection pool for each postgres instance
2. When `REPLICA_CONFIG` comes in it picks up a random replica checks if it's eligible for the query and uses the connection for the query (or transaction) or uses a primary instead
3. Replica state is cached, when next similar request comes in EdgeDB can choose a connection based on the cache instead of checking a random replica

Read-Only EdgeDB Instance
-------------------------

Read-only replicas of Postgres allow scaling database performance, but read-only EdgeDB Instances are needed for two reasons:

1. If single instance's IO is saturated
2. If primary is unavailable (i.e. a primary and a replica is in different DC and it makes sense to colocate read-only EdgeDB with the replica, so that if cross-DC link is down, data is still accessible in read-only mode)

After the proposed changes in the protocol the only challenge left is to synchronize schema between normal EdgeDB instance and read-only ones.

Protocol
========

`Prepare` message gets the following header:

* `REPLICA_CONFIG` = `{}` -- this means request can go to the replica and must be readonly. This is a JSON field. It's an object and settings like maximum replication lag of the replica will be added later (clarification: empty object `{}` means any read-only replica can be used). The list of settings is open-ended. Unknown setting means "always use primary".

When `REPLICA_CONFIG` is not specified and transaction is active, the config specified in `START TRANSACTION` is used (because transaction requires all requests being in the same connection). Specifying different config in the middle of transaction is an error.

`PrepareComplete` message gets the following header:
* `READ_ONLY` = `true`/`false` -- marks read-only or read-write requests

New error types:

* `QueryIsNotReadOnly` when read-write query sent with `REPLICA_CONFIG`
* `InconsistentReplicaConfig` when replica config of the `START TRANSACTION` is different than current query and transaction is still active.

Open Questions
==============

**Should credentials file contain a list of replicas?**

In the current spec we specify a different credentials file for read-only replica EdgeDB. I think in most cases there will be no replica EdgeDB, and all EdgeDB instance will be connected to both primary and replicas. And setup with `read-only` EdgeDB is quite rare.

**This proposal allows accessing replicas on the same connection**

The alternative is to always open separate connection to replica. But the issue with that is that sometimes we need different replica settings: i.e.  ensure that replica lag is less that 10 seconds. To make that working we need to poll all connections to find out if there is one such replica. On the other hand, EdgeDB is able to do that more efficiently.

This implementation is signicantly harder, but we don't have to implement all of that at start, we just have to specify it in a way that allows us to implement that feature later.

**REPLICA_CONFIG might be disallowed in the middle of transaction**

This Proposal allows sending the header but requires keeping it the same. We may as well disallow the header in the middle of transaction.

Either way this should probably be disallowed on the API level (i.e. `Transaction` object doesn't have `read_only` method) rather than reaching EdgeDB.

**Unknown setting in REPLICA_CONFIG might mean error**

This way we have to increase protocol version or negotiate extension to use additional constaint on replica. Which is also a good option.

The Plan
========

* 1.0 should have `read_only` API and stage (1) of failover of Postgres
* Subsequent updates (in rought order, but may be changed depending due to demand):
    * Replica read by EdgeDB
    * Read-only EdgeDB instance
    * Replica selection settings
    * Stage (2) of failover
* And the following two things can probably be in the commercial version only:
    * Scaling EdgeDB processes (i.e. multiple writer processes)
    * Setting up replication of the postgres and election of primary by the
    EdgeDB itself (failover stage (3))

**Comment by elprans** [comment]
> Established connections are not refreshed, because it's not possible to reliably determine whether new IP address resolved by DNS is because of failover or it's just DNS round-robin returned a different IP.

Not sure I understand this bit.  What do you mean by "established connection" here and why DNS resolution would be a concern for an open connection?

> Read-Only Requests and Replicas

Aside from the explicit API prescribed here, we should also consider routing read-only queries to replicas automatically without a need for an explicit `.read_only()`.  In a synchronous replication setup replica lag is not a concern and micro-managing `.read_only()` might quickly become very tedious.  This is especially useful for `pool.query()` and the like where you don't even get to see an explicit connection object.

> `conn = await edgedb.async_connection_pool(warn={'non_transactional_writes'})`

It is perfectly fine to do a mutation in `pool.query()` without an explicit transaction, no need to warn about that.  I imagine that lots of application API handlers could be written as a single `INSERT` or `UPDATE` query on a pool.

> empty object `{}` means empty replica can be used

What do you mean by "empty replica" here?

> `REPLICA_CONFIG` might be disallowed in the middle of transaction

I think it's fine to keep it, but check consistency.  Otherwise we might make some forms of composition unnecessarily difficult (clients having to check if in transaction or not before forming messges etc.).

Good start overall, thank you @tailhook!

**Reply by tailhook** [reply]
> > Established connections are not refreshed, because it's not possible to reliably determine whether new IP address resolved by DNS is because of failover or it's just DNS round-robin returned a different IP.
> 
> Not sure I understand this bit. What do you mean by "established connection" here and why DNS resolution would be a concern for an open connection?
> 

1. If `mydb.example.org` refers to `10.0.0.1` and we have 10 connections to it
2. We reconnect one, and  `mydb.example.org` already refers to `10.0.0.2`
3. No we have 1 connection to `10.0.0.2` and 9 connections to `10.0.0.1`

So there are two cases when IP can be different:
1. EdgeDB is being migrated to a different IP, all connections should eventually move to `10.0.0.2`
2. EdgeDB has two IP addresses and DNS round-robin is used for load-balancing between them

Both are valid use cases, but by stating the text above we cover use case (2) and rely that (1) is handled by server (i.e. server will close connections when it expects that clients should migrate to a different address). Since we don't have a way for closing connection gracefully from server this only works because we rely on client retrying all requests on network failure.

> > Read-Only Requests and Replicas
> 
> Aside from the explicit API prescribed here, we should also consider routing read-only queries to replicas automatically without a need for an explicit `.read_only()`. In a synchronous replication setup replica lag is not a concern and micro-managing `.read_only()` might quickly become very tedious. This is especially useful for `pool.query()` and the like where you don't even get to see an explicit connection object.
> 

Yes, connection pool option should enable this. We probably may do that by default only if we detect synchronous replication.

> > `conn = await edgedb.async_connection_pool(warn={'non_transactional_writes'})`
> 
> It is perfectly fine to do a mutation in `pool.query()` without an explicit transaction, no need to warn about that. I imagine that lots of application API handlers could be written as a single `INSERT` or `UPDATE` query on a pool.
> 

But we can't reliably retry them, which leads to errors when server migrates between hosts or restarts. We can make it explicit like:
```
conn.query(..., retry=True)  # or
conn.query(..., transaction=True)
```

> > empty object `{}` means empty replica can be used
> 
> What do you mean by "empty replica" here?
> 

"any read-only replica can be used" (some weird typo)

> > `REPLICA_CONFIG` might be disallowed in the middle of transaction
> 
> I think it's fine to keep it, but check consistency. Otherwise we might make some forms of composition unnecessarily difficult (clients having to check if in transaction or not before forming messges etc.).
> 

This was also my thought.

**Comment by dmgolembiowski** [comment]
> # Open Questions
> **Should credentials file contain a list of replicas?**
> 
> In the current spec we specify a different credentials file for read-only replica EdgeDB. I think in most cases there will be no replica EdgeDB, and all EdgeDB instance will be connected to both primary and replicas. And setup with `read-only` EdgeDB is quite rare.

After some light reading on [Agiledata.org](http://www.agiledata.org/essays/mappingObjects.html), I'm thinking it's a good idea to optionally include a list of replicas in the credentials file based upon:
>> 5.2 Lazy Reads
An important performance consideration is whether the attribute should be automatically read in when the object is retrieved. When an attribute is very large, for example the picture of a person could be 100k whereas the rest of the attributes are less than 1k, and rarely accessed you may want to consider taking a lazy read approach. The basic idea is that instead of automatically bringing the attribute across the network when the object is read you instead retrieve it only when the attribute is actually needed. This can be accomplished by a getter method, an operation whose purpose is to provide the value of a single attribute, that checks to see if the attribute has been initialized and if not retrieves it from the database at that point.
Other common uses for lazy read is reporting and for retrieving objects as the results of searches where you only need a small subset of the data of an object.

**Reply by tailhook** [reply]
Well, this reminded be that stating a question like that is a mistake: we don't have list of hosts in credentials file (just one host). And agreed that complex host configuration will be made programmatically anyway (the discussioin is here: https://github.com/edgedb/edgedb/discussions/1683)
